keps/sig-api-machinery/2309-idl/README.md

```diff
@@ -0,0 +1,636 @@
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+### Graduation Criteria
+
+## Implementation History
+
+## Alternatives
+
+### OpenAPI
+
+* **Hard to write by hand**: OpenAPI is a great output format for
+  validation, but writing by hand is difficult.  For instance, in order to
+  describe a tagged union (a common construct in Kubernetes: consider
+  horizontal pod autoscalers, the update strategy in workload controllers,
+  etc) in Kubernetes-compatible OpenAPI, a complicated construct involving
+  oneof, single-value enums, and property-count validation is necessary. 
+
+* **Doesn’t describe all of the Kubernetes type system**: we already have
+  a bunch of custom extensions on top of OpenAPI (e.g. x-int-or-string),
+  which at least partially negates the benefit that comes from shared
+  OpenAPI tooling.
+
+* **Describes invalid types**: OpenAPI can easily describe constructs
+  which are either invalid (e.g. certain combinations of oneof under
+  structural rules) or not recommended (e.g. complex maps, floats).  Many
+  of these are tools that users reach for instinctively, and as such
+  OpenAPI guides users down the wrong path (see goal #1)
+
+* **Not a particularly good code generation source**: for similar reasons
+  to “hard to write by hand”, OpenAPI does not actually make
+  a particularly good code generation source for type-safe APIs.  For
+  instance:
+
+  * Existing tooling is pretty good at generating “naive” types from
+    OpenAPI, it’s hard to extract meaning on more complex types (e.g.
+    unions, enums).
+
+  * It’s currently impossible to identify stringly-typed types more
+    robustly in CRD OpenAPI due to type erasure.
+
+    <details>
+    <summary>stringly typed values...</summary>
+
+    ... are values whose concrete type is string, but which have
+    structured data inside the string.  Examples in k8s include Quantity,
+    Time, and Duration.
+
+    </details>
+
+  * Extensions could improve these cases, but that means:
+
+    * maintaining custom tooling to work with those custom extensions (not
+      that different from maintaining tooling that ingests IDL IR)
+
+    * Potentially waiting for API changes & API server version support
+      (often up to 1y after release), unless you want to maintain an extra
+      translation layer for older API servers.
+
+### Proto
```

Comment:
Note that the proto for kube is the only IDL we have, just a custom one.  It is a full representation of all complexity in k/k and at least a few other projects.  The logic in the current generator is representative of the scope of work to bring k/k into an IDL.  The additional logic to generate go types from that IDL is what go-to-protobuf already does.

So I would say this section is not really correct - whether proto is our desired language or not, it is the only IDL we have today that is complete.

Reply:
What do you mean by "complete" & "IDL" here?  I think we may have different definitions.

For me,

- "complete" means "comprehensively and accurately captures our type system".  While the proto files certainly accurately capture a subset, they're far from comprehensive -- they lack higher-level concepts like kinds, list-maps, etc, and certainly don't capture validation or defaulting

- "IDL" to me implies a source of truth for your type system.  In this case, then, our proto files certainly don't meet that criterion: We *generate* our proto files, so they can't be a source of truth.  In fact, they're more-or-less just an intermediate artifact, since we discard they types they generate entirely.

All that being said, perhaps I've miscommunicated the intent of this section -- it's not supposed to be "why not adopt our current files as the source of truth", it's "why not use proto syntax & the proto compiler to make a source of truth".

Reply:
Proto files != IDL.  We have a generated IDL from go struct that is required to express all structural capability AND round trip back to the same go types, so while there are elements that don't cover needs in our type system, the current conversion (and all its complexity) has to be adapted to any new thing.

Reply:
I'm really failing to parse your words here, sorry.  Are you just saying that Go types that are generated from KDL have to be equivalent to Go types generated by the proto compiler?

If so, yeah, I agree, that's certainly the case ATM, that's why I'm focusing on the Go types produced by KDL being identical to those we have currently.  We could theoretically revisit that decision in the future (in light of gogoproto disappearing, since IIRC we depend heavily on extensions from that?), but I feel like that'd be hella work and pretty complicated, so I've no intention to try that now.